physics/checkpointer_body.hpp

```diff
@@ -0,0 +1,75 @@
+
+#pragma once
+
+#include "physics/checkpointer.hpp"
+
+namespace principia {
+namespace physics {
+namespace internal_checkpointer {
+
+template<typename Message>
+Checkpointer<Message>::Checkpointer(Reader reader, Writer writer)
+    : reader_(std::move(reader)),
+      writer_(std::move(writer)) {}
+
+template<typename Message>
+void Checkpointer<Message>::CreateUnconditionally(Instant const& t) {
+  absl::MutexLock l(&lock_);
+  CreateUnconditionallyLocked(t);
+}
+
+template<typename Message>
+bool Checkpointer<Message>::CreateIfNeeded(
+    Instant const& t,
+    Time const& max_time_between_checkpoints) {
+  absl::MutexLock l(&lock_);
+  if (checkpoints_.empty() ||
+      max_time_between_checkpoints < t - checkpoints_.crbegin()->first) {
+    CreateUnconditionallyLocked(t);
+    return true;
+  }
+  return false;
+}
+
+template<typename Message>
+void Checkpointer<Message>::ForgetBefore(Instant const& t) {
+  absl::MutexLock l(&lock_);
+  auto const it = checkpoints_.lower_bound(t);
+  CHECK(it == checkpoints_.end() || t <= it->first);
+  checkpoints_.erase(checkpoints_.begin(), it);
+}
+
+template<typename Message>
+Instant Checkpointer<Message>::WriteToMessage(
+    not_null<Message*> const message) const {
+  absl::ReaderMutexLock l(&lock_);
+  if (checkpoints_.empty()) {
+    // TODO(phl): declare this next to Instant.
+    static Instant infinite_future = Instant() + quantities::Infinity<Time>();
+    return infinite_future;
+  } else {
+    message->MergeFrom(checkpoints_.cbegin()->second);
```

Comment:
Will this have the correct behaviour for repeated fields, since those are appended instead of overwritten? If it does not, there should be a comment warning of that in the header file.

Reply:
Added a comment.